File: python/day2/prb.py
import itertools

# with open("../../inputs/d2") as f:
with open("../../inputs/d2") as f:
    lines = f.readlines()

ans = 0


def is_safe(nums):
    diffs = [b - a for a, b in itertools.zip_longest(nums, nums[1:]) if b is not None]
    return all([a >= 1 and a <= 3 for a in diffs]) or all(
        [a <= -1 and a >= -3 for a in diffs]
    )


def is_safe_p2(nums):
    if is_safe(nums):
        return True

    for n in range(0, len(nums)):
        tmp = nums.copy()
        tmp.pop(n)
        if is_safe(tmp):
            return True
    return False


ans2 = 0

for line in lines:
    nums = [int(entry) for entry in line.strip().split()]
    if is_safe(nums):
        ans += 1
    if is_safe_p2(nums):
        ans2 += 1


assert ans == 299


assert ans2 == 364

File: typescript/day2/prb.ts
const content = await Deno.readTextFile("../../inputs/d2");

const tmp = content.split("\n").filter((line: string) => line.trim());
let lines = tmp.map((value: string) =>
  value.split(" ").map((value: string) => Number(value))
);

const is_safe = (x: number[]): boolean => {
  let n = x.length;
  let diffs: number[] = [];
  for (let i: number = 1; i < n; i++) {
    diffs.push(x[i] - x[i - 1]);
  }
  const asc = diffs.every((value) => value >= 1 && value <= 3);
  const desc = diffs.every((value: number): boolean =>
    value >= -3 && value <= -1
  );
  return (asc || desc) && x.length > 0;
};

const ans = lines.filter(is_safe);

const check_part2 = (line: number[]): boolean => {
  if (is_safe(line)) return true;

  let len = line.length;
  for (let x: number = 0; x <= len; x++) {
    if (is_safe(line.filter((_e, i) => i != x))) return true;
  }

  return false;
};

const ans2 = lines.filter(check_part2);

console.log(`Part1: ${ans.length}`);
console.log(`Part2: ${ans2.length}`);
